File: app/src/main/java/com/mmi/sdk/demo/java/activity/AutoSuggestActivity.java
package com.mmi.sdk.demo.java.activity;


import android.os.Bundle;
import android.os.Handler;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import com.mapbox.mapboxsdk.annotations.MarkerOptions;
import com.mapbox.mapboxsdk.camera.CameraUpdateFactory;
import com.mapbox.mapboxsdk.geometry.LatLng;
import com.mapbox.mapboxsdk.maps.MapboxMap;
import com.mapbox.mapboxsdk.maps.OnMapReadyCallback;
import com.mapbox.mapboxsdk.maps.SupportMapFragment;
import com.mmi.sdk.demo.R;
import com.mmi.sdk.demo.java.adapter.AutoSuggestAdapter;
import com.mmi.sdk.demo.java.utils.CheckInternet;
import com.mmi.sdk.demo.java.utils.TransparentProgressDialog;
import com.mmi.services.api.auth.MapmyIndiaAuthentication;
import com.mmi.services.api.auth.model.AtlasAuthToken;
import com.mmi.services.api.autosuggest.MapmyIndiaAutoSuggest;
import com.mmi.services.api.autosuggest.model.AutoSuggestAtlasResponse;
import com.mmi.services.api.autosuggest.model.ELocation;
import com.mmi.services.api.geocoding.GeoCode;
import com.mmi.services.api.geocoding.GeoCodeResponse;
import com.mmi.services.api.geocoding.MapmyIndiaGeoCoding;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * Created by CEINFO on 26-02-2019.
 */

public class AutoSuggestActivity extends AppCompatActivity implements OnMapReadyCallback, TextWatcher {

    private MapboxMap mapboxMap;
    private EditText autoSuggestText;
    private RecyclerView recyclerView;
    private LinearLayoutManager mLayoutManager;
    private TransparentProgressDialog transparentProgressDialog;
    private Handler handler;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.autosuggest_activity);

        SupportMapFragment mapFragment =
                (SupportMapFragment) getSupportFragmentManager().findFragmentById(R.id.mapBoxId);
        mapFragment.getMapAsync(this);

        initReferences();
        initListeners();
    }

    private void initListeners() {
        autoSuggestText.addTextChangedListener(this);
    }

    private void initReferences() {
        autoSuggestText = findViewById(R.id.auto_suggest);
        recyclerView = findViewById(R.id.recyclerview);
        mLayoutManager = new LinearLayoutManager(AutoSuggestActivity.this);
        recyclerView.setLayoutManager(mLayoutManager);
        recyclerView.setVisibility(View.GONE);

        if (CheckInternet.isNetworkAvailable(AutoSuggestActivity.this)) {
            getAuthToken();
        } else {
            showToast(getString(R.string.pleaseCheckInternetConnection));
        }

        transparentProgressDialog = new TransparentProgressDialog(this, R.drawable.circle_loader, "");
        handler = new Handler();
    }

    @Override
    public void onMapReady(MapboxMap mapboxMap) {
        this.mapboxMap = mapboxMap;

        mapboxMap.setMinZoomPreference(4.5);
        mapboxMap.setMaxZoomPreference(18.5);


        mapboxMap.setPadding(20, 20, 20, 20);


        mapboxMap.moveCamera(CameraUpdateFactory.newLatLngZoom(new LatLng(
                22.553147478403194,
                77.23388671875), 4));
    }

    @Override
    public void onMapError(int i, String s) {
    }

    public void selectedPlaceName(String name) {
        if (CheckInternet.isNetworkAvailable(AutoSuggestActivity.this)) {
            getGeoCode(name);
        } else {
            showToast(getString(R.string.pleaseCheckInternetConnection));
        }
    }

    private void getGeoCode(String geocodeText) {
        show();
        new MapmyIndiaGeoCoding.Builder()
                .setAddress(geocodeText)
                .build().enqueueCall(new Callback<GeoCodeResponse>() {
            @Override
            public void onResponse(Call<GeoCodeResponse> call, Response<GeoCodeResponse> response) {
                if (response.code() == 200) {
                    if (response.body() != null) {
                        List<GeoCode> placesList = response.body().getResults();
                        GeoCode place = placesList.get(0);
                        String add = "Latitude: " + place.latitude + " longitude: " + place.longitude;
                        addMarker(place.latitude, place.longitude);
                        showToast(add);
                    } else {
                        showToast("Not able to get value, Try again.");
                    }
                } else {
                    showToast(response.message());
                }
                hide();
            }

            @Override
            public void onFailure(Call<GeoCodeResponse> call, Throwable t) {
                showToast(t.toString());
                hide();
            }
        });
    }

    private void addMarker(double latitude, double longitude) {
        mapboxMap.clear();
        mapboxMap.addMarker(new MarkerOptions().position(new LatLng(
                latitude, longitude)));
    }


    private void callAutoSuggestApi(String searchString) {
        new MapmyIndiaAutoSuggest.Builder()
                .setQuery(searchString)
                .build()
                .enqueueCall(new Callback<AutoSuggestAtlasResponse>() {
                    @Override
                    public void onResponse(Call<AutoSuggestAtlasResponse> call, Response<AutoSuggestAtlasResponse> response) {
                        if (response.code() == 200) {
                            if (response.body() != null) {
                                ArrayList<ELocation> suggestedList = response.body().getSuggestedLocations();
                                if (suggestedList.size() > 0) {
                                    recyclerView.setVisibility(View.VISIBLE);
                                    AutoSuggestAdapter autoSuggestAdapter = new AutoSuggestAdapter(suggestedList, name -> {
                                        selectedPlaceName(name);
                                        recyclerView.setVisibility(View.GONE);
                                    });
                                    recyclerView.setAdapter(autoSuggestAdapter);
                                }
                            } else {
                                showToast("Not able to get value, Try again.");
                            }
                        }
                    }

                    @Override
                    public void onFailure(Call<AutoSuggestAtlasResponse> call, Throwable t) {
                        showToast(t.toString());
                    }
                });

    }

    private void show() {
        transparentProgressDialog.show();
    }

    private void hide() {
        transparentProgressDialog.dismiss();
    }

    private void showToast(String msg) {
        Toast.makeText(AutoSuggestActivity.this, msg, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {

    }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {
        handler.postDelayed(() -> {
            recyclerView.setVisibility(View.GONE);
            if (s.length() < 3)
                recyclerView.setAdapter(null);

            if (s != null && s.toString().trim().length() < 2) {
                recyclerView.setAdapter(null);
                return;
            }

            if (s.length() > 2) {
                if (CheckInternet.isNetworkAvailable(AutoSuggestActivity.this)) {
                    callAutoSuggestApi(s.toString());
                } else {
                    showToast(getString(R.string.pleaseCheckInternetConnection));
                }
            }
        }, 300);

    }

    @Override
    public void afterTextChanged(Editable s) {
    }

    private void getAuthToken() {

        new MapmyIndiaAuthentication.Builder().build().enqueueCall(new Callback<AtlasAuthToken>() {
            @Override
            public void onResponse(Call<AtlasAuthToken> call, Response<AtlasAuthToken> response) {
            }

            @Override
            public void onFailure(Call<AtlasAuthToken> call, Throwable t) {

            }
        });

    }

}

File: app/src/main/java/com/mmi/sdk/demo/kotlin/activity/AutoSuggestActivity.kt
package com.mmi.sdk.demo.kotlin.activity

import android.os.Bundle
import android.os.Handler
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.EditText
import android.widget.Toast
import com.mapbox.mapboxsdk.annotations.MarkerOptions
import com.mapbox.mapboxsdk.camera.CameraUpdateFactory
import com.mapbox.mapboxsdk.geometry.LatLng
import com.mapbox.mapboxsdk.maps.MapboxMap
import com.mapbox.mapboxsdk.maps.OnMapReadyCallback
import com.mapbox.mapboxsdk.maps.SupportMapFragment
import com.mmi.sdk.demo.R
import com.mmi.sdk.demo.java.utils.CheckInternet
import com.mmi.sdk.demo.kotlin.adapter.AutoSuggestAdapter
import com.mmi.sdk.demo.kotlin.kotlin.utility.TransparentProgressDialog
import com.mmi.services.api.auth.MapmyIndiaAuthentication
import com.mmi.services.api.auth.model.AtlasAuthToken
import com.mmi.services.api.autosuggest.MapmyIndiaAutoSuggest
import com.mmi.services.api.autosuggest.model.AutoSuggestAtlasResponse
import com.mmi.services.api.geocoding.GeoCodeResponse
import com.mmi.services.api.geocoding.MapmyIndiaGeoCoding
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

/**
 * Created by CEINFO on 26-02-2019.
 */
class AutoSuggestActivity : AppCompatActivity(), OnMapReadyCallback, TextWatcher {

    lateinit var mapboxMap: MapboxMap
    private lateinit var autoSuggestText: EditText
    lateinit var recyclerView: RecyclerView
    lateinit var handler: Handler
    private var mLayoutManager: LinearLayoutManager? = null
    private lateinit var transparentDialog: TransparentProgressDialog
    lateinit var authToken: String
    lateinit var tokenType: String

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.autosuggest_activity)
        val mapFragment = supportFragmentManager
                .findFragmentById(R.id.mapBoxId) as SupportMapFragment?
        mapFragment!!.getMapAsync(this)

        initReference()
    }

    override fun onMapReady(mapboxMap: MapboxMap?) {
        this.mapboxMap = mapboxMap!!
        mapboxMap.setMinZoomPreference(4.5)
        mapboxMap.setMaxZoomPreference(18.5)

        mapboxMap.setPadding(20, 20, 20, 20)
        mapboxMap.moveCamera(CameraUpdateFactory.newLatLngZoom(LatLng(
                22.553147478403194,
                77.23388671875), 4.0))
    }

    override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
        handler.postDelayed(Runnable {
            recyclerView.visibility = View.GONE
            if (s!!.length < 3)
                recyclerView.adapter = null

            if (s != null && s.toString().trim { it <= ' ' }.length < 2) {
                recyclerView.adapter = null
                return@Runnable
            }

            if (s.length > 2) {
                if (CheckInternet.isNetworkAvailable(this@AutoSuggestActivity)) {
                    callAutoSuggestApi(s.toString())
                } else {
                    showToast(getString(R.string.pleaseCheckInternetConnection))
                }
            }
        }, 300)
    }

    override fun onMapError(p0: Int, p1: String?) {}

    override fun afterTextChanged(s: Editable?) {}

    override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}

    private fun initReference() {
        recyclerView = findViewById(R.id.recyclerview)
        autoSuggestText = findViewById(R.id.auto_suggest)

        mLayoutManager = LinearLayoutManager(this@AutoSuggestActivity)
        recyclerView.layoutManager = mLayoutManager
        recyclerView.visibility = View.GONE

        if (CheckInternet.isNetworkAvailable(this@AutoSuggestActivity)) {
            getAuthToken()
        } else {
            showToast(getString(R.string.pleaseCheckInternetConnection))
        }

        handler = Handler()
        autoSuggestText.addTextChangedListener(this)
    }

    fun selectedPlaceName(name: String?) {
        if (name != null) {
            if (CheckInternet.isNetworkAvailable(this@AutoSuggestActivity)) {
                getGeoCode(name)
            } else {
                showToast(getString(R.string.pleaseCheckInternetConnection))
            }

        }
    }

    private fun getGeoCode(geocodeText: String) {
        show()
        MapmyIndiaGeoCoding.Builder<MapmyIndiaGeoCoding.Builder<*>>()
                .setAddress(geocodeText)
                .build().enqueueCall(object : Callback<GeoCodeResponse> {
                    override fun onResponse(call: Call<GeoCodeResponse>, response: Response<GeoCodeResponse>) {
                        if (response.code() == 200) {
                            if (response.body() != null) {
                                val placesList = response.body()?.results
                                val place = placesList?.get(0)
                                addMarker(place?.latitude!!, place?.longitude!!)
                            } else {
                                showToast("Not able to get value, Try again.")
                                response.message()
                            }
                        } else {
                            showToast(response.message())
                        }
                        hide()
                    }

                    override fun onFailure(call: Call<GeoCodeResponse>, t: Throwable) {
                        showToast(t.toString())
                        hide()
                    }
                })
    }

    private fun addMarker(latitude: Double, longitude: Double) {
        mapboxMap.clear()
        mapboxMap.addMarker(MarkerOptions().position(LatLng(
                latitude, longitude)))
    }

    @Synchronized
    private fun callAutoSuggestApi(searchString: String) {

        MapmyIndiaAutoSuggest.Builder<MapmyIndiaAutoSuggest.Builder<*>>()
                .setQuery(searchString)
                .build()
                .enqueueCall(object : Callback<AutoSuggestAtlasResponse> {
                    override fun onResponse(call: Call<AutoSuggestAtlasResponse>, response: Response<AutoSuggestAtlasResponse>) {
                        if (response.code() == 200) {
                            if (response.body() != null) {
                                val suggestedList = response.body()!!.suggestedLocations
                                if (suggestedList.size > 0) {
                                    recyclerView.visibility = View.VISIBLE
                                    val autoSuggestAdapter = AutoSuggestAdapter(suggestedList, object : AutoSuggestAdapter.PlaceName {
                                        override fun nameOfPlace(name: String) {
                                            selectedPlaceName(name)
                                            recyclerView.visibility = View.GONE
                                        }
                                    })
                                    recyclerView.adapter = autoSuggestAdapter
                                }
                            } else {
                                Toast.makeText(this@AutoSuggestActivity, "Not able to get value, Try again.", Toast.LENGTH_SHORT).show()
                            }
                        }
                    }

                    override fun onFailure(call: Call<AutoSuggestAtlasResponse>, t: Throwable) {
                        println(t.toString())
                    }
                })
    }

    private fun getAuthToken() {

        MapmyIndiaAuthentication.Builder<MapmyIndiaAuthentication.Builder<*>>()
                .build()
                .enqueueCall(object : Callback<AtlasAuthToken> {
                    override fun onResponse(call: Call<AtlasAuthToken>, response: Response<AtlasAuthToken>) {
                        if (response.code() == 200) {
                            authToken = response.body()!!.getAccessToken()
                            tokenType = response.body()!!.getTokenType()
                        }

                    }

                    override fun onFailure(call: Call<AtlasAuthToken>, t: Throwable) {

                    }
                })
    }

    fun show() {
        transparentDialog = TransparentProgressDialog(this, R.drawable.circle_loader, "")
        transparentDialog.show()
    }

    fun hide() {
        transparentDialog.dismiss()
    }

    fun showToast(message: String) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }

}